cc/mcts_player.cc

```diff
@@ -0,0 +1,247 @@
+
+    std::cout << root_->position.ToPrettyString();
+    std::cout << root_->Describe();
+
+    if (ShouldResign()) {
+      SetResult(root_->position.to_play() == Color::kBlack ? -1 : 1,
+                root_->position.CalculateScore(),
+                GameOverReason::kOpponentResigned);
+      break;
+    }
+
+    auto move = PickMove();
+    PlayMove(move);
+    if (root_->position.is_game_over()) {
+      float score = root_->position.CalculateScore();
+      float result = score < 0 ? -1 : score > 0 ? 1 : 0;
+      SetResult(result, score, GameOverReason::kBothPassed);
+      break;
+    } else if (root_->position.n() >= kMaxSearchDepth) {
+      float score = root_->position.CalculateScore();
+      float result = score < 0 ? -1 : score > 0 ? 1 : 0;
+      SetResult(result, score, GameOverReason::kMoveLimitReached);
+      break;
+    }
+
+    std::cout << "Q: " << std::setw(8) << std::setprecision(5) << root_->Q()
+              << "\n";
+    std::cout << "Played >>" << move << std::endl;
+  }
+}
+
+Coord MctsPlayer::PickMove() {
+  if (!options_.soft_pick || root_->position.n() > temperature_cutoff_) {
+    // Choose the most visited node.
+    Coord c = ArgMax(root_->edges, MctsNode::CmpN);
+    std::cout << "Picked arg_max " << c << "\n";
+    return c;
+  }
+
+  // Select from the first kN * kN moves (instead of kNumMoves) to avoid
+  // randomly choosing to pass early on in the game.
+  std::array<float, kN * kN> cdf;
+
+  cdf[0] = root_->child_N(0);
+  for (size_t i = 1; i < cdf.size(); ++i) {
+    cdf[i] = cdf[i - 1] + root_->child_N(i);
+  }
+  float norm = 1 / cdf[cdf.size() - 1];
+  for (size_t i = 0; i < cdf.size(); ++i) {
+    cdf[i] *= norm;
+  }
+  float e = rnd_();
+  Coord c = SearchSorted(cdf, e);
+  std::cout << "Picked rnd(" << e << ") " << c << "\n";
+  assert(root_->child_N(c) != 0);
+  return c;
+}
+
+void MctsPlayer::TreeSearch(int batch_size) {
+  int max_iterations = batch_size * 2;
+
+  // TODO(tommadams): Avoid creating this vector each time.
+  std::vector<MctsNode*> leaves;
+  for (int i = 0; i < max_iterations; ++i) {
+    auto* leaf = root_->SelectLeaf();
+    if (leaf == nullptr) {
+      continue;
+    }
+    if (leaf->position.is_game_over() ||
+        leaf->position.n() >= kMaxSearchDepth) {
+      float value = leaf->position.CalculateScore() > 0 ? 1 : -1;
+      leaf->IncorporateEndGameResult(value, root_);
+    } else {
+      leaf->AddVirtualLoss(root_);
+      leaves.push_back(leaf);
+      if (static_cast<int>(leaves.size()) == batch_size) {
+        break;
+      }
+    }
+  }
+
+  if (!leaves.empty()) {
+    // TODO(tommadams): Avoid creating these vectors each time.
+    std::vector<const DualNet::BoardFeatures*> features;
+    features.reserve(leaves.size());
+    for (auto* leaf : leaves) {
+      features.push_back(&leaf->features);
+    }
+
+    std::vector<DualNet::Output> outputs;
+    outputs.resize(leaves.size());
+
+    RunMany(features, {outputs.data(), outputs.size()});
```

Comment:
Incidentally, this may need some re-work for TPUs, which may require a constant batch-size for speed.